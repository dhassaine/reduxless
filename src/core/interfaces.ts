/**
 * The initial state can be provided as either a plain JavaScript object,
 * a Map of string keys and values, or an array of arrays of string keys
 * and values to generate the initial state from.
 *
 * Here's an example of a plain object:
 * ```js
 * { counter: 0, updatedAt: Date.now() }
 * ```
 *
 * Here's a map of key, value pairs:
 * ```js
 * new Map([
 *   ['counter', 0],
 *   ['updatedAt': Date.now()]
 * ])
 * ```
 *
 * Here's a doubly-nested array of key, value pairs:
 * ```js
 * [
 *   ['counter', 0],
 *   ['updatedAt', Date.now()]
 * ]
 * ```
 */
export type EnumerableObject<T> =
  | { [index: string]: T }
  | Map<string, T>
  | [string, T][];

/**
 * A validator for a given mount point in the store will receive a value and
 * return a `true` or `false` if the value is valid or not.
 *
 * If you would like validation to throw an error, this is where you can opt
 * into doing it.
 */
export type Validator = (value: any) => boolean;

export type Validators = EnumerableObject<Validator>;

interface CreateStoreArgs {
  /** pairs of keys (mountpoints) and data */
  initialState?: EnumerableObject<any>;
  /** pairs of mountpoints and validator functions */
  validators?: Validators;
  /**
   * An optional scheduling function that is passed a pending store update that
   * needs to be invoked.
   *
   * For example, to batch store updates on every `requestAnimationFrame`, you
   * can pass a function that will collect pending updates to be executed in
   * batches according to the frame scheduler:
   *
   * ```js
   * let _rafId = null;
   * let pendingUpdates = [];
   *
   * const store = createStore({
   *   batchUpdateFn: (pendingStoreUpdate) => {
   *     if (_rafId) cancelAnimationFrame();
   *     pendingUpdates.push(pendingStoreUpdate);
   *     _rafId = requestAnimationFrame(() => {
   *       _rafId = null;
   *       pendingUpdates.forEach(update => update());
   *       pendingUpdates = [];
   *     });
   *   }
   * })
   * ```
   */
  batchUpdateFn?: (fn: () => void) => any;
}

/** Creates a simple store instance. */
export type CreateStore = (args?: CreateStoreArgs) => Store;

export interface Serializer {
  toUrlValue: (value: any) => string;
  fromUrlValue: (value: string) => any;
}

export type Serializers = Map<string, Serializer>;

interface CreateRouterEnabledStoreArgs extends CreateStoreArgs {
  pushStateMountPoints?: string[];
  replaceStateMountPoints?: string[];
  serializers?: EnumerableObject<Serializer>;
  routerOptions?: {
    debounceTime?: number;
    useHash?: boolean;
  };
}

/**
 * Creates a router-enabled store instance that syncs the browser URL with
 * select store data.
 */
export type CreateRouterEnabledStore = (
  args?: CreateRouterEnabledStoreArgs
) => RouterEnabledStore;

type MountPoint = string;

interface MountPointsToValues {
  [keys: string]: any;
}

type Unsubscribe = () => void;

/** The store instance */
export interface Store {
  /**
   * To modify the store, you can call `set()` with the key or mountpoint
   * in the store and the data to replace the mountpoint with. After the
   * `set` function is executed all the subscribers to the store will be
   * notified.
   *
   * ```js
   * store.set("name", "Homer");
   * ```
   */
  set: (mountPoint: MountPoint, value: any) => void;
  /**
   * Use this if you wish to modify multiple mountpoints simultaneously
   * and have a single notification emit to the subscribers.
   *
   * ```js
   * store.setAll({ name: "Homer", score: 1 });
   * ```
   */
  setAll: (mountPointsToValues: EnumerableObject<any>) => void;
  /**
   * To retrieve a single property from the store use `get` with the appropriate
   * mountpoint.
   *
   * ```js
   * store.get("name");
   *  ```
   */
  get: (mountPoint: MountPoint) => any;
  /**
   * To retrieve multiple properties from the store use `getAll` with an array
   * containing all the desired mountpoints.
   *
   * ```js
   * store.getAll(["name", "score"]); // returns {name: 'Homer', score: 1}
   * ```
   */
  getAll: (mountPoints: MountPoint[]) => MountPointsToValues;
  /**
   * This function allows you to control the update phase of the store with more
   * precision. The function argument is called with a store containing the
   * setter and getter functions. The subscribers are only notified once after
   * the `fn` argument has executed. Any calls to `set` or `setAll` in `fn` will
   * not cause an unnecessary subscriber notification.
   *
   * A good example use case is where you wish to set two properties at the same
   * time, but one property relies on a projection of the other:
   *
   * ```js
   * store.withMutations(s => {
   *   const oldTop = selectors.top(store);
   *   const oldCellHeight = selectors.cellHeight(store);
   *   s.set("isZoomedIn", !!zoom);
   *   s.set("top", (oldTop / oldCellHeight) * selectors.cellHeight(store));
   * });
   * ```
   */
  withMutations: (fn: (mutableStore: Store) => void) => void;
  /**
   * This registers a function that will be called on every state update before
   * the observers are notified.
   */
  addUpdateIntercept: (fn: () => any) => void;
  /**
   * Notifies the subscribers to the store, even when no changes occured to the
   * store.
   */
  ping: () => void;
  /**
   * Adds `fn` to a list of observers that will be executed every time the store
   * is updated. It returns an `unsubscribe` function so you can remove the
   * observer later.
   */
  subscribe: (fn: (storeApi: Store) => any) => Unsubscribe;
}

/** The router-enabled store instance */
export interface RouterEnabledStore extends Store {
  syncToLocations: string[];
  useHash: boolean;
  syncedLocationToStore: boolean;
  serializers: Serializers;
  navigate: (newPath?: string) => void;
}

/**
 * A selector function that uses the values in the store and the props on the
 * container component to generate a computed value.
 */
export type SelectorMapper = (store: Store, ownProps: any) => any;

export type SelectorMappings = EnumerableObject<SelectorMapper>;

/**
 * A function that performs modifications to the store or side effects outside
 * of the store.
 */
export type ActionMapper = (store: Store, ownProps: any, ...args: any[]) => any;

export type ActionMappings = EnumerableObject<ActionMapper>;
