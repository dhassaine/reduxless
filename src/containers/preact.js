import { Component } from 'preact';
import { createContainer, createMapper } from './factory';

export const Container = createContainer(Component);
export const mapper = createMapper(Component);
