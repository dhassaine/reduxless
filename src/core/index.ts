import { createRouterEnabledStore } from './router';
import createStore from './state/store';
import selectorMemoizer from './utilities/memoizer';

export { createStore, selectorMemoizer, createRouterEnabledStore };
