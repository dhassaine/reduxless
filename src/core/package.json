{
  "name": "@reduxless/core",
  "version": "5.0.1",
  "description": "A small and performant state management and routing library with unidirectional data flow",
  "author": "Djamel Hassaine <dhassaine@gmail.com> (https://github.com/dhassaine)",
  "contributors": [
    "Klemen Slaviƒç <krof.drakula@gmail.com> (https://github.com/KrofDrakula)"
  ],
  "license": "MIT",
  "main": "index.js",
  "module": "index.esm.js",
  "minified:main": "index.min.js",
  "types": "./types",
  "jest": {
    "testEnvironment": "jsdom",
    "setupFiles": [
      "<rootDir>/test-setup/polyfill.ts"
    ],
    "moduleDirectories": [
      "node_modules"
    ],
    "moduleFileExtensions": [
      "js",
      "jsx",
      "ts",
      "tsx"
    ],
    "transform": {
      "\\.(j|t)sx?$": "ts-jest"
    },
    "testURL": "http://example.com/page1?queryParam=queryValue&a[]=1&a[]=2&storeData=%7B%22counter%22%3A%7B%22value%22%3A1%7D%2C%22counter2%22%3A%7B%22value%22%3A2%7D%7D"
  }
}
