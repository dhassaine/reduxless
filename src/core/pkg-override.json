{
  "name": "@reduxless/core",
  "description": "A small and performant state management and routing library with unidirectional data flow",
  "main": "index.js",
  "module": "index.esm.js",
  "minified:main": "index.min.js",
  "types": "."
}
