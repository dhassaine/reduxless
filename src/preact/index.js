import Wrapper from "../index";

const {
  Container,
  mapper,
  Link,
  Match,
  createStore,
  enableHistory,
  selectorMemoizer
} = Wrapper(require("preact"));

export {
  Container,
  mapper,
  Link,
  Match,
  createStore,
  enableHistory,
  selectorMemoizer
};
