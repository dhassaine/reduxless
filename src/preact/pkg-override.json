{
  "name": "@reduxless/preact",
  "description": "Preact bindings for @reduxless/core",
  "main": "index.js",
  "module": "index.esm.js",
  "minified:main": "index.min.js",
  "types": ".",
  "peerDependencies": {
    "@reduxless/core": "*",
    "preact": ">=10"
  }
}
