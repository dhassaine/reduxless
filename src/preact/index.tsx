/** @jsx h */
/** @jsxFrag Fragment */
import {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  h,
  createContext,
  FunctionalComponent,
  Component,
  JSX,
  Fragment,
} from 'preact';
import { getPath } from '@reduxless/core';
import type {
  SelectorMappings,
  ActionMappings,
  Store,
  RouterEnabledStore,
} from '@reduxless/core/interfaces';

const StoreContext = createContext<{ store: Store | RouterEnabledStore }>(null);

export const Container: FunctionalComponent<{
  store: Store | RouterEnabledStore;
}> = ({ store, children }) => (
  <StoreContext.Provider value={{ store }}>{children}</StoreContext.Provider>
);

export const mapper =
  (propMappings: SelectorMappings = {}, actionMappings: ActionMappings = {}) =>
  (Wrapped: FunctionalComponent) => {
    return class extends Component<{ store?: Store | RouterEnabledStore }> {
      _unsubscribe: () => void;
      _mappedProps: SelectorMappings = null;
      _mappedActions: ActionMappings = null;
      _store: Store | RouterEnabledStore = null;

      shouldComponentUpdate(nextProps) {
        for (const key in nextProps) {
          if (this.props[key] !== nextProps[key]) return true;
        }
        return false;
      }

      componentDidMount() {
        if (Object.keys(propMappings).length !== 0)
          this._unsubscribe = this._store.subscribe(this.handleStoreUpdate);
      }

      componentWillUnmount() {
        if (this._unsubscribe) {
          this._unsubscribe();
          this._unsubscribe = null;
        }
      }

      handleStoreUpdate = () => {
        let hasPropsChanged = false;
        for (const key of Object.keys(propMappings)) {
          const prop = propMappings[key](this._store, this.props);
          if (prop !== this._mappedProps[key]) hasPropsChanged = true;
          this._mappedProps[key] = prop;
        }

        if (hasPropsChanged) this.forceUpdate();
      };

      render() {
        const { children, store, ...rest } = this.props;
        return (
          <StoreContext.Consumer>
            {({ store }) => {
              if (!this._store) this._store = store;
              if (!this._mappedProps) {
                this._mappedProps = {};
                for (const key of Object.keys(propMappings)) {
                  this._mappedProps[key] = propMappings[key](
                    this._store,
                    this.props
                  );
                }
              }
              if (!this._mappedActions) {
                this._mappedActions = {};
                for (const key of Object.keys(actionMappings)) {
                  this._mappedActions[key] = (...args) =>
                    actionMappings[key](this._store, this.props, ...args);
                }
              }
              return (
                <Wrapped
                  {...rest}
                  {...this._mappedProps}
                  {...this._mappedActions}
                >
                  {children}
                </Wrapped>
              );
            }}
          </StoreContext.Consumer>
        );
      }
    };
  };

export const Link: FunctionalComponent<JSX.HTMLAttributes<HTMLAnchorElement>> =
  ({ href, children, ...rest }) => (
    <StoreContext.Consumer>
      {({ store }) => (
        <a
          {...rest}
          href={href}
          onClick={(ev) => {
            ev.preventDefault();
            (store as RouterEnabledStore).navigate(href);
          }}
        >
          {children}
        </a>
      )}
    </StoreContext.Consumer>
  );

interface MatchProps {
  path: string | ((path: string) => boolean);
  currentPath?: string;
}

export const Match: FunctionalComponent<MatchProps> = ({
  path,
  currentPath,
  children,
}) => {
  return (
    <StoreContext.Consumer>
      {({ store }) => {
        const actualPath = currentPath ?? getPath(store as RouterEnabledStore);
        const matched =
          typeof path == 'function'
            ? path(actualPath)
            : actualPath.split('?')[0] == path;
        return matched ? <Fragment>{children}</Fragment> : null;
      }}
    </StoreContext.Consumer>
  );
};
