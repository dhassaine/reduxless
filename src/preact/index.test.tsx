/** @jsx h */

import {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  h,
} from 'preact';
import { render } from '@testing-library/preact';

import { Container, Match } from './index';
import { createRouterEnabledStore } from '@reduxless/core';

describe('Smoke test', () => {
  it('Match renders children if url path matches', async () => {
    const store = createRouterEnabledStore();
    const results = render(
      <Container store={store}>
        <Match path={'/page1'}>
          <div>hello</div>
        </Match>
        <Match path={'/page2'}>
          <div>bye</div>
        </Match>
        <Match path={(p) => p == '/page3'}>
          <div>bye</div>
        </Match>
      </Container>,
    );
    expect(results.asFragment()).toMatchSnapshot();
  });
});
