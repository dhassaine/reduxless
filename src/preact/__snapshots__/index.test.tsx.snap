// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Smoke test Match renders children if url path matches 1`] = `
<DocumentFragment>
  <div>
    hello
  </div>
</DocumentFragment>
`;
